File: desktop/ai-app-bridge-cli/bin/script/script-sdk.js
'use strict';

const fs = require('node:fs');
const readline = require('node:readline');

const artifactPath = process.argv[2];
const replies = new Map();
const controlWaiters = [];
let nextId = 0;
let control = { status: 'running', pauseReason: null };

const rl = readline.createInterface({ input: process.stdin });

function send(message) {
  fs.writeSync(1, `${JSON.stringify(message)}\n`);
}

function writeStderr(args) {
  fs.writeSync(2, `${args.map(String).join(' ')}\n`);
}

console.log = (...args) => writeStderr(args);
console.info = (...args) => writeStderr(args);
console.debug = (...args) => writeStderr(args);

function waitReply(id) {
  return new Promise((resolve) => {
    if (replies.size >= 32) {
      send({ type: 'fail', error: 'channel_backlog' });
      process.exit(1);
      return;
    }
    replies.set(id, resolve);
  });
}

rl.on('line', (line) => {
  if (!line) return;
  let message;
  try {
    message = JSON.parse(line);
  } catch {
    send({ type: 'fail', error: 'malformed_frame' });
    process.exit(1);
    return;
  }
  if (message.type === 'start') {
    applyControl(message.control);
    run(message).catch((error) => {
      send({ type: 'fail', error: error.message || String(error) });
      process.exit(1);
    });
    return;
  }
  if (message.type === 'control') {
    applyControl(message.control);
    return;
  }
  if (message.id && replies.has(message.id)) {
    applyControl(message.control);
    const resolve = replies.get(message.id);
    replies.delete(message.id);
    resolve(message);
  }
});

send({ type: 'ready' });

function applyControl(next) {
  if (!next) return;
  control = next;
  while (controlWaiters.length > 0) controlWaiters.shift()();
}

function isHeld(status) {
  return status === 'pause_requested'
    || status === 'paused_manual'
    || status === 'paused_live'
    || status === 'paused_ambiguous';
}

function holdIfPaused() {
  return new Promise((resolve) => {
    const tick = () => {
      if (!isHeld(control.status)) {
        resolve();
        return;
      }
      controlWaiters.push(tick);
    };
    tick();
  });
}

async function run(start) {
  const loaded = require(artifactPath);
  const entry = start.entrypoint || 'main';
  const main = typeof loaded === 'function' && entry === 'main' ? loaded : loaded[entry];
  if (typeof main !== 'function') {
    send({ type: 'fail', error: 'unsupported_entrypoint' });
    process.exit(1);
  }
  const ctx = {
    inputs: start.inputs || {},
    call: async (command, args, options) => {
      await holdIfPaused();
      const id = `c${nextId += 1}`;
      send({ type: 'call', id, command, args: args || {}, options: options || {} });
      const reply = await waitReply(id);
      await holdIfPaused();
      return reply.value;
    },
    assert: async (assertion) => {
      await holdIfPaused();
      const id = `a${nextId += 1}`;
      send({ type: 'assert', id, assertion });
      const reply = await waitReply(id);
      await holdIfPaused();
      return reply.value;
    },
    progress: async (event) => {
      await holdIfPaused();
      send({ type: 'progress', event });
      await holdIfPaused();
    },
    checkpoint: async (name, state) => {
      await holdIfPaused();
      const id = `k${nextId += 1}`;
      send({ type: 'checkpoint', id, name, state });
      const reply = await waitReply(id);
      await holdIfPaused();
      return reply.value;
    },
    askAgent: async (request) => {
      await holdIfPaused();
      const id = `q${nextId += 1}`;
      send({ type: 'ask', id, request });
      const reply = await waitReply(id);
      await holdIfPaused();
      return reply.value;
    },
    controlPoint: () => control,
    resume: () => {
      const state = control || {};
      if (!state.checkpoint) return null;
      return state.checkpoint.state;
    },
  };
  const result = await main(ctx);
  send({ type: 'return', result });
  process.exit(0);
}

File: desktop/ai-app-bridge-cli/bin/script/script-sdk.py
import asyncio
import json
import sys


_real_stdout = sys.stdout
HELD = {
    "pause_requested",
    "paused_manual",
    "paused_live",
    "paused_ambiguous",
}


def send(message):
    _real_stdout.write(json.dumps(message) + "\n")
    _real_stdout.flush()


def read_message():
    line = sys.stdin.readline()
    if not line:
        return None
    return json.loads(line)


class HostContext:
    def __init__(self, inputs, control):
        self.inputs = inputs
        self._control = control
        self._next_id = 0

    def _id(self, prefix):
        self._next_id += 1
        return f"{prefix}{self._next_id}"

    def _apply_control(self, payload):
        if payload:
            self._control = payload

    def _hold_if_paused(self):
        while self._control.get("status") in HELD:
            message = read_message()
            if message is None:
                raise RuntimeError("channel_closed")
            if message.get("type") == "control":
                self._apply_control(message.get("control"))
                continue
            raise RuntimeError("unexpected_frame")

    def _read_reply(self, message_id):
        while True:
            reply = read_message()
            if reply is None:
                raise RuntimeError("channel_closed")
            if reply.get("type") == "control":
                self._apply_control(reply.get("control"))
                continue
            if reply.get("id") == message_id:
                self._apply_control(reply.get("control"))
                return reply
            raise RuntimeError("unexpected_frame")

    def _sync_control(self):
        message_id = self._id("s")
        send({"type": "control-point", "id": message_id})
        self._read_reply(message_id)

    def _safe_point(self):
        self._sync_control()
        self._hold_if_paused()

    def call(self, command, args=None, options=None):
        self._safe_point()
        message_id = self._id("c")
        send({"type": "call", "id": message_id, "command": command, "args": args or {}, "options": options or {}})
        reply = self._read_reply(message_id)
        self._safe_point()
        return reply["value"]

    def assert_(self, assertion):
        self._safe_point()
        message_id = self._id("a")
        send({"type": "assert", "id": message_id, "assertion": assertion})
        reply = self._read_reply(message_id)
        self._safe_point()
        return reply["value"]

    def progress(self, event):
        self._safe_point()
        send({"type": "progress", "event": event})
        self._safe_point()

    def checkpoint(self, name, state):
        self._safe_point()
        message_id = self._id("k")
        send({"type": "checkpoint", "id": message_id, "name": name, "state": state})
        reply = self._read_reply(message_id)
        self._safe_point()
        return reply["value"]

    def askAgent(self, request):
        self._safe_point()
        message_id = self._id("q")
        send({"type": "ask", "id": message_id, "request": request})
        reply = self._read_reply(message_id)
        self._safe_point()
        return reply["value"]

    def controlPoint(self):
        self._sync_control()
        return self._control

    def resume(self):
        checkpoint = self._control.get("checkpoint")
        if not checkpoint:
            return None
        return checkpoint["state"]


def load_namespace(path):
    namespace = {}
    with open(path, "r", encoding="utf-8") as handle:
        exec(handle.read(), namespace, namespace)
    return namespace


def main():
    artifact_path = sys.argv[1]
    send({"type": "ready"})
    start = read_message()
    sys.stdout = sys.stderr
    ctx = HostContext(start.get("inputs") or {}, start.get("control") or {"status": "running", "pauseReason": None})
    entry = start.get("entrypoint") or "main"
    try:
        namespace = load_namespace(artifact_path)
        loaded = namespace.get(entry)
        if not callable(loaded):
            send({"type": "fail", "error": "unsupported_entrypoint"})
            sys.exit(1)
        result = loaded(ctx)
        if asyncio.iscoroutine(result):
            result = asyncio.run(result)
        send({"type": "return", "result": result})
    except Exception as error:
        send({"type": "fail", "error": str(error)})
        sys.exit(1)


if __name__ == "__main__":
    main()
